## Cleaning and analyzing *Used-car* Sales data
#### Description
The dataset is a derivation from the German eBay website used car sales data analysis.

The goal of this project is to clean, analyze and identify relevant data.

In [1]:
import pandas as pd
import numpy as np

In [2]:
autos = pd.read_csv('autos.csv',encoding='Latin-1')

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [4]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

##### Our dataset contains 20 columns, most of which are stored as strings.

In [5]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

##### Here we understand that our data does have null values but they do not account for more than 20% of a particular column.

In [6]:
autos.describe()

yearOfRegistration       powerPS  monthOfRegistration  nrOfPictures  \
count        50000.000000  50000.000000         50000.000000       50000.0   
mean          2005.073280    116.355920             5.723360           0.0   
std            105.712813    209.216627             3.711984           0.0   
min           1000.000000      0.000000             0.000000           0.0   
25%           1999.000000     70.000000             3.000000           0.0   
50%           2003.000000    105.000000             6.000000           0.0   
75%           2008.000000    150.000000             9.000000           0.0   
max           9999.000000  17700.000000            12.000000           0.0   

         postalCode  
count  50000.000000  
mean   50813.627300  
std    25779.747957  
min     1067.000000  
25%    30451.000000  
50%    49577.000000  
75%    71540.000000  
max    99998.000000

#### The data above shows 50000 count of year of registration of such ads with an average of 2005 as the year of registration and a minimum of 1000 and maximum of 9999 which is evidently a flaw in the data.
Most of the data is invisible for insights due to it's string nature/dtype.
Hence, we'll have to clean the data first.

In [7]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

##### Most of columns being in Camelcase is a little difficult to process from the Python preferred snakecase.
Hence, we'll convert/clean the column names first for easy working with the access of such columns.

## Cleaning column-names
***Starting with 'yearOfRegistration' to 'registration_year','monthOfRegistration' to 'registration_month','notRepairedDamage' to 'unrepaired_damage' and 'date_created' to 'ad_created' and so on and forth.***

In [8]:
autos.columns =  ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']


In [9]:
autos.head(2)

date_crawled                                        name  seller  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   

  offer_type   price  ab_test vehicle_type  registration_year    gearbox  \
0    Angebot  $5,000  control          bus               2004    manuell   
1    Angebot  $8,500  control    limousine               1997  automatik   

   power_ps   model   odometer  registration_month fuel_type    brand  \
0       158  andere  150,000km                   3       lpg  peugeot   
1       286     7er  150,000km                   6    benzin      bmw   

  unrepaired_damage           ad_created  num_photos  postal_code  \
0              nein  2016-03-26 00:00:00           0        79588   
1              nein  2016-04-04 00:00:00           0        71034   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08

### There!
##### We have the new column names in the snakecase representation.

### *Now working with these columns shall be easier.*
*The current range majorly corresponds to the ease of accessing these columns but it is likewise an important step for reducing redundant work.*

## Exploration and further bits of cleaning

In [10]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-14 20:50:02  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

##### In the above description, we come across a few redundant value holding columns and also a couple of anomalies in a few of the columns.
*To further the investigation, we'll have to take a closer look at a few of the columns to understand better of their values.*

In [11]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        50000 non-null  object
 1   name                50000 non-null  object
 2   seller              50000 non-null  object
 3   offer_type          50000 non-null  object
 4   price               50000 non-null  object
 5   ab_test             50000 non-null  object
 6   vehicle_type        44905 non-null  object
 7   registration_year   50000 non-null  int64 
 8   gearbox             47320 non-null  object
 9   power_ps            50000 non-null  int64 
 10  model               47242 non-null  object
 11  odometer            50000 non-null  object
 12  registration_month  50000 non-null  int64 
 13  fuel_type           45518 non-null  object
 14  brand               50000 non-null  object
 15  unrepaired_damage   40171 non-null  object
 16  ad_created          50

In [12]:
autos[['seller','price','offer_type','ab_test','odometer','num_photos','unrepaired_damage']].head(90)

seller   price offer_type  ab_test   odometer  num_photos  \
0   privat  $5,000    Angebot  control  150,000km           0   
1   privat  $8,500    Angebot  control  150,000km           0   
2   privat  $8,990    Angebot     test   70,000km           0   
3   privat  $4,350    Angebot  control   70,000km           0   
4   privat  $1,350    Angebot     test  150,000km           0   
..     ...     ...        ...      ...        ...         ...   
85  privat  $5,000    Angebot     test   60,000km           0   
86  privat    $399    Angebot  control  125,000km           0   
87  privat      $0    Angebot  control  150,000km           0   
88  privat  $1,800    Angebot     test  150,000km           0   
89  privat  $7,900    Angebot  control  125,000km           0   

   unrepaired_damage  
0               nein  
1               nein  
2               nein  
3               nein  
4               nein  
..               ...  
85               NaN  
86                ja  
87               NaN  
88              nein  
89              nein  

[90 rows x 7 columns]

### Derivations from investigation
*The above sample print of the suspected columns reveals some interesting data.*

- columns 'seller', 'offer_type' have redundant data that under assumptions of analysis are unactionable, _hence_ worth dropping.
- column 'unrepaired_damage' has data that is relevant to user inference.
- column 'num_photos' only has 0 value data, showcasing redundancy again, _hence_ worth dropping.
- column 'a/b test' has actionable relevance to website management.
- columns 'price' and 'odometer' have numeric data with text association resulting in text storage of its values, _hence_ a conversion to numeric types is essential.

### Derivations result:
> ***Drop 'seller', 'offer_type' and 'num_photos'.***

> ***Convert 'price' and 'odometer' to numeric types and modify the names of the columns likewise.***

In [13]:
autos = autos.drop(['offer_type', 'seller', 'num_photos'], axis=1)

In [14]:
autos.head(2)

date_crawled                                        name   price  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD  $5,000   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  $8,500   

   ab_test vehicle_type  registration_year    gearbox  power_ps   model  \
0  control          bus               2004    manuell       158  andere   
1  control    limousine               1997  automatik       286     7er   

    odometer  registration_month fuel_type    brand unrepaired_damage  \
0  150,000km                   3       lpg  peugeot              nein   
1  150,000km                   6    benzin      bmw              nein   

            ad_created  postal_code            last_seen  
0  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1  2016-04-04 00:00:00        71034  2016-04-06 14:45:08

#### DROPPED.

In [15]:
autos.info()
#autos['odometer'].info()
autos['price'] = (autos['price'].str.replace("$","").str.replace(",","").astype(float))
autos['odometer'] = (autos['odometer'].str.replace("km","") .str.replace(",","").astype(int))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        50000 non-null  object
 1   name                50000 non-null  object
 2   price               50000 non-null  object
 3   ab_test             50000 non-null  object
 4   vehicle_type        44905 non-null  object
 5   registration_year   50000 non-null  int64 
 6   gearbox             47320 non-null  object
 7   power_ps            50000 non-null  int64 
 8   model               47242 non-null  object
 9   odometer            50000 non-null  object
 10  registration_month  50000 non-null  int64 
 11  fuel_type           45518 non-null  object
 12  brand               50000 non-null  object
 13  unrepaired_damage   40171 non-null  object
 14  ad_created          50000 non-null  object
 15  postal_code         50000 non-null  int64 
 16  last_seen           50

In [16]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   date_crawled        50000 non-null  object 
 1   name                50000 non-null  object 
 2   price               50000 non-null  float64
 3   ab_test             50000 non-null  object 
 4   vehicle_type        44905 non-null  object 
 5   registration_year   50000 non-null  int64  
 6   gearbox             47320 non-null  object 
 7   power_ps            50000 non-null  int64  
 8   model               47242 non-null  object 
 9   odometer            50000 non-null  int64  
 10  registration_month  50000 non-null  int64  
 11  fuel_type           45518 non-null  object 
 12  brand               50000 non-null  object 
 13  unrepaired_damage   40171 non-null  object 
 14  ad_created          50000 non-null  object 
 15  postal_code         50000 non-null  int64  
 16  last

In [17]:
autos.rename({'odometer': 'odometer_km'}, axis=1, inplace=True)

#### CONVERTED AND MODIFIED.

## EDA Start:

Columns that show anomalies need further exploration.
_Therefore_, start by exploring the numeric columns of the dataframe.

In [27]:
autos['price'].unique().shape

(2357,)

In [29]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

#### Here we can spot the anomaly. The minimum value of any car cannot be 0$.

In [32]:
autos['price'].value_counts().head(50)

0.0       1421
500.0      781
1500.0     734
2500.0     643
1200.0     639
1000.0     639
600.0      531
800.0      498
3500.0     498
2000.0     460
999.0      434
750.0      433
900.0      420
650.0      419
850.0      410
700.0      395
4500.0     394
300.0      384
2200.0     382
950.0      379
1100.0     376
1300.0     371
3000.0     365
550.0      356
1800.0     355
5500.0     340
1250.0     335
350.0      335
1600.0     327
1999.0     322
400.0      321
6500.0     300
1400.0     292
2800.0     291
250.0      291
2300.0     290
1350.0     276
1700.0     268
200.0      266
450.0      265
3800.0     264
3200.0     261
2900.0     256
1450.0     252
4000.0     246
2999.0     242
2600.0     241
5000.0     239
1900.0     239
3900.0     238
Name: price, dtype: int64

In [33]:
autos['odometer_km'].unique().shape

(13,)

In [34]:
autos['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [38]:
autos['odometer_km'].value_counts().head(13)

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

*Now that we've assessed the unique values, their counts and frequencies, we can investigate further with understanding the anomalies of the 'price' column.* 

__THEREFORE, WE'LL TAKE A LOOK AT THE MAXIMUM AND MINIMUM VALUES OF THE 'price' column to assess the situation.__

In [40]:
autos['price'].value_counts().sort_index(ascending = True).head(50)

0.0      1421
1.0       156
2.0         3
3.0         1
5.0         2
8.0         1
9.0         1
10.0        7
11.0        2
12.0        3
13.0        2
14.0        1
15.0        2
17.0        3
18.0        1
20.0        4
25.0        5
29.0        1
30.0        7
35.0        1
40.0        6
45.0        4
47.0        1
49.0        4
50.0       49
55.0        2
59.0        1
60.0        9
65.0        5
66.0        1
70.0       10
75.0        5
79.0        1
80.0       15
89.0        1
90.0        5
99.0       19
100.0     134
110.0       3
111.0       2
115.0       2
117.0       1
120.0      39
122.0       1
125.0       8
129.0       1
130.0      15
135.0       1
139.0       1
140.0       9
Name: price, dtype: int64

**Observation : The price of 0 USD has 1421 listings, roughly 3 percent of the total listings.**
*Additionally, 1 USD bids are present as well. These bids can be justified due to the nature of the website being an auction site.*

*__But__ the 0$ listings are invalid.*

Trying the opposite side for exploration, starting from the maximum values.

In [41]:
autos['price'].value_counts().sort_index(ascending = False).head(50)

99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
350000.0      1
345000.0      1
299000.0      1
295000.0      1
265000.0      1
259000.0      1
250000.0      1
220000.0      1
198000.0      1
197000.0      1
194000.0      1
190000.0      1
180000.0      1
175000.0      1
169999.0      1
169000.0      1
163991.0      1
163500.0      1
155000.0      1
151990.0      1
145000.0      1
139997.0      1
137999.0      1
135000.0      1
130000.0      1
129000.0      1
128000.0      1
120000.0      2
119900.0      1
119500.0      1
116000.0      1
115991.0      1
115000.0      1
114400.0      1
109999.0      1
105000.0      2
104900.0      1
99900.0       2
99000.0       2
98500.0       1
Name: price, dtype: int64

On the maximum side, the values seem to stabilize by 130,000 USD, hence consideration of data upto 130,000 USD is necessary.

In [46]:
autos = autos[autos["price"].between(1,131000)]
autos["price"].describe()

count     48541.000000
mean       5789.902989
std        7764.188436
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7450.000000
max      130000.000000
Name: price, dtype: float64

## EDA on the date columns for understanding the date range covers.

#### The 'date_crawled', 'last_seen', and 'ad_created' columns are all identified as string values.
#### 'registration_year' and 'registration_month' are in the form of integers.

Understanding the format of the values present in the string columns of date-like data.

In [47]:
autos[["date_crawled","ad_created","last_seen"]][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

The date is presented in the form of YYYY-MM-DD.
_Therefore, we can try extracting the first few characters by accessing the specific columns._

In [48]:
print(autos['date_crawled'].str[:10])

0        2016-03-26
1        2016-04-04
2        2016-03-26
3        2016-03-12
4        2016-04-01
            ...    
49995    2016-03-27
49996    2016-03-28
49997    2016-04-02
49998    2016-03-08
49999    2016-03-14
Name: date_crawled, Length: 48541, dtype: object


In [51]:
autos["date_crawled"].str[:10].value_counts(normalize=True, dropna=False).sort_values(ascending = True)
#method chaining to get the distribution alongside the value counts of the date format in the date_crawled column.
#ascending order of date format data frequency distribution is implemented with the sort values function.

2016-04-07    0.001401
2016-04-06    0.003173
2016-03-18    0.012896
2016-04-05    0.013082
2016-03-06    0.014050
2016-03-13    0.015677
2016-03-05    0.025298
2016-03-24    0.029336
2016-03-16    0.029604
2016-03-27    0.031067
2016-03-25    0.031602
2016-03-17    0.031623
2016-03-31    0.031829
2016-03-10    0.032200
2016-03-26    0.032220
2016-03-23    0.032220
2016-03-11    0.032591
2016-03-22    0.032962
2016-03-09    0.033106
2016-03-08    0.033291
2016-03-30    0.033662
2016-04-01    0.033683
2016-03-29    0.034115
2016-03-15    0.034301
2016-03-19    0.034795
2016-03-28    0.034837
2016-04-02    0.035496
2016-03-07    0.036031
2016-04-04    0.036505
2016-03-14    0.036567
2016-03-12    0.036938
2016-03-21    0.037350
2016-03-20    0.037885
2016-04-03    0.038607
Name: date_crawled, dtype: float64

In [52]:
autos["registration_year"].describe()

count    48541.000000
mean      2004.753260
std         88.665531
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

##### HERE AN ANOMALY IS EASILY DETECTED!
__The minimum and maximum values of the 'registration_year' are 1000 and 9999 respectively. This confirms of possibility of wrongly entered data, therefore, a correction is necessary.__

_Since no car registration years corresponding to values higher than the year of 2016 can be presented, values above 2016 need to be invalidated._

_Also, cars were invented in the 1900s, therefore, listings with years prior to 1900s can be invalidated._

In [55]:
autos = autos[autos["registration_year"].between(1900,2016)]

### Normalizing the brands popularity.

In [61]:
autos["brand"].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [62]:
autos["brand"].value_counts(normalize = True)

volkswagen        0.211372
bmw               0.110037
opel              0.107637
mercedes_benz     0.096470
audi              0.086589
ford              0.069936
renault           0.047174
peugeot           0.029856
fiat              0.025655
seat              0.018282
skoda             0.016418
nissan            0.015282
mazda             0.015196
smart             0.014167
citroen           0.014017
toyota            0.012710
hyundai           0.010031
sonstige_autos    0.009773
volvo             0.009152
mini              0.008766
mitsubishi        0.008230
honda             0.007844
kia               0.007073
alfa_romeo        0.006644
suzuki            0.005937
porsche           0.005787
chevrolet         0.005701
chrysler          0.003515
dacia             0.002636
daihatsu          0.002508
jeep              0.002272
subaru            0.002143
land_rover        0.002100
saab              0.001650
jaguar            0.001565
daewoo            0.001500
trabant           0.001393
r

Since the values_count() function gives the distribution of the respective brands on the website, we can select the top 20 as a sample set to understand the effects of brand values in the current dataset.

The top 20 brands also represent data above 5% dominance of the current listings in terms of the count.

In [67]:
brands_dist = autos["brand"].value_counts(normalize = True)

In [68]:
popular_brands = brands_dist[brands_dist > .05].index

In [69]:
popular_brands

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')

As evident in the above index series, the most popular brands are volkswagen, bmw, opel,mercedes-benz, audi and ford.

_Hence these brands can be appropriately utilized to work upon in aggregation._


### Calculating the mean price.

In [78]:
pop_brands_mean_price = {}

for b in popular_brands:
    autos_pop_brands = autos[autos["brand"] == b]
    autos_pop_brands_mean_price = autos_pop_brands["price"].mean()
    pop_brands_mean_price[b] = float(autos_pop_brands_mean_price)
pop_brands_mean_price

{'volkswagen': 5402.410261610221,
 'bmw': 8201.538566419946,
 'opel': 2975.2419354838707,
 'mercedes_benz': 8554.524105754277,
 'audi': 9295.681683168317,
 'ford': 3749.4695065890287}

__The above data helps in realizing that popular brands by prices are in the order of:__

__audi > mercedes-benz > bmw > volkswagen > ford > opel__

__HIGHEST > HIGHER       > HIGH >    MEDIUM >  LOW > LOWEST__

### Calculating the mean mileage.

*Here the pop_brands_mean_price can be utilized to form a series object of the mileage data from the 'odometer_km' column.*

In [80]:
bmp_series = pd.Series(pop_brands_mean_price)
print(bmp_series)

volkswagen       5402.410262
bmw              8201.538566
opel             2975.241935
mercedes_benz    8554.524106
audi             9295.681683
ford             3749.469507
dtype: float64


In [81]:
df = pd.DataFrame(bmp_series, columns=['mean_price'])
df

mean_price
volkswagen     5402.410262
bmw            8201.538566
opel           2975.241935
mercedes_benz  8554.524106
audi           9295.681683
ford           3749.469507

In [86]:
pop_brands_mean_mileage = {}
for b in popular_brands:
    pop_brands_mileage = autos[autos["brand"] == b]
    autos_pop_brands_mean_mileage = pop_brands_mileage["odometer_km"].mean()
    pop_brands_mean_mileage[b] = float(autos_pop_brands_mean_mileage)
pop_brands_mean_mileage

{'volkswagen': 128707.15879132022,
 'bmw': 132633.4242306194,
 'opel': 129310.0358422939,
 'mercedes_benz': 130826.48300377694,
 'audi': 129188.11881188118,
 'ford': 124266.01287159056}

In [89]:
bmm_series = pd.Series(pop_brands_mean_mileage)

In [90]:
bmm_series

volkswagen       128707.158791
bmw              132633.424231
opel             129310.035842
mercedes_benz    130826.483004
audi             129188.118812
ford             124266.012872
dtype: float64

In [91]:
df["mean mileage"] = bmm_series

In [94]:
df.sort_values("mean_price", ascending = True)

mean_price   mean mileage
opel           2975.241935  129310.035842
ford           3749.469507  124266.012872
volkswagen     5402.410262  128707.158791
bmw            8201.538566  132633.424231
mercedes_benz  8554.524106  130826.483004
audi           9295.681683  129188.118812

_There is no correlation between the mean price and mean mileage for the respective dataset._